hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieAppendHandle.java

```diff
@@ -189,8 +191,14 @@ protected boolean isUpdateRecord(HoodieRecord<T> hoodieRecord) {
   private Option<IndexedRecord> getIndexedRecord(HoodieRecord<T> hoodieRecord) {
     Option<Map<String, String>> recordMetadata = hoodieRecord.getData().getMetadata();
     try {
-      Option<IndexedRecord> avroRecord = hoodieRecord.getData().getInsertValue(tableSchema,
-          config.getProps());
+      // Pass the isUpdateRecord to the props for HoodieRecordPayload to judge
+      // Whether it is a update or insert record.
+      boolean isUpdateRecord = isUpdateRecord(hoodieRecord);
```

Comment:
HoodieAppendHandle should handle update records, here you mean use global index?

Reply:
Here I just pass the `isUpdateRecord` flag to the `ExpressionPayload`. So it can know current record is a matched record or not matched record. The matched record will execute the match-clause in merge-into, while the not-matched record will execute the not-match-clause. If we do not have such information, the result of merge into will be incorrect.